app/src/main/java/org/jellyfin/androidtv/ui/startup/LoginViewModel.kt

```diff
@@ -0,0 +1,108 @@
+package org.jellyfin.androidtv.ui.startup
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.liveData
+import androidx.lifecycle.switchMap
+import org.jellyfin.androidtv.data.model.LoadingState
+import org.jellyfin.androidtv.data.model.Server
+import org.jellyfin.androidtv.data.model.ServerList
+import org.jellyfin.androidtv.data.model.User
+import org.jellyfin.androidtv.data.repository.ServerRepository
+import org.jellyfin.androidtv.data.repository.UserRepository
+import org.jellyfin.androidtv.util.apiclient.toUser
+import timber.log.Timber
+
+class LoginViewModel(
+	private val serverRepository: ServerRepository,
+	private val userRepository: UserRepository
+) : ViewModel() {
+	private val _currentServer = MutableLiveData<Server>()
+	val currentServer: LiveData<Server>
+		get() = _currentServer
+
+	private val _currentUser = MutableLiveData<User>()
+	val currentUser: LiveData<User>
+		get() = _currentUser
+
+	private val _serverList = MediatorLiveData<ServerList>()
+	val serverList: LiveData<ServerList>
+		get() = _serverList
+
+	private val _loadingCurrentServer = MutableLiveData<LoadingState>()
+	private val currentServerUsers = _currentServer.switchMap { server ->
+		liveData {
+			_loadingCurrentServer.value = LoadingState.LOADING
+			emit(hashMapOf(server to userRepository.getUsers(server)))
+			_loadingCurrentServer.value = LoadingState.SUCCESS
+		}
+	}
+
+	private val _loadingDiscoveredServers = MutableLiveData<LoadingState>()
+	private val discoveredServers = liveData {
+		_loadingDiscoveredServers.value = LoadingState.LOADING
+		emit(serverRepository.discoverServers()
+				 .associateWith { userRepository.getUsers(it) })
+		_loadingDiscoveredServers.value = LoadingState.SUCCESS
+	}
+
+	private val _loadingSavedServers = MutableLiveData<LoadingState>()
+	private val savedServers = liveData {
+		_loadingSavedServers.value = LoadingState.LOADING
+		emit(serverRepository.getServers()
+				 .associateWith { userRepository.getUsers(it) })
+		_loadingSavedServers.value = LoadingState.SUCCESS
+	}
+
+	init {
+		// Add all loading states
```

Comment:
Using something like
```kt
_serverList.apply {
    addSource(_loadingCurrentServer) {				
```

will probably make it more readable, just add all addSource calls in the .apply {} block